### PR TITLE
Fixes bugs with the number with unit form

### PR DESCRIPTION
**Pull request by ryan-bradford**
## PR Checklist

Please check if your PR fulfills the following requirements:

-   [X] Tests for the changes have been added (for bug fixes / features)
-   [X] Examples have been added / updated (for bug fixes / features)
-   [X] Changelog has been updated

## PR Type

What kind of change does this PR introduce?

-   [X] Bugfix
-   [ ] Feature
-   [ ] Code style update (formatting, local variables)
-   [ ] Refactoring (no functional changes, no api changes)
-   [ ] Build related changes
-   [ ] CI related changes
-   [ ] Documentation content changes
-   [ ] Example website changes
-   [ ] Version bump
-   [ ] Other... Please describe:

## What does this change do?

1. Fixes a bug where if the unlimitedValue was 0, it was impossible to set the form as required. This is because null was passed through in `getValue()` outputting 0, which was the same as the unlimtedValue.
2. Fixes a bug where if the default value was unlimited, a unit was never calculated causing errors. 
3. Changes it so if the user types in the unlimited value, when the input is blurred, the checkbox will be checked.

## What manual testing did you do?
1. Typed in the unlimited value, blurred the input, observed the checkbox get checked
2. Set the unlimitedValue to 0 and observed a required error when I cleared the form.
3. Set the default to unlimited through the form control and observed a calculated unit on unselected the unlimited box.

## GIf

number-with-unit

## Does this PR introduce a breaking change?

-   [ ] Yes
-   [X] No

### CODE REVIEW

**Comment by bryanvb** [comment]
Looks good to me.

File: projects/components/src/form/number-with-unit-input/number-with-unit-form-input.component.spec.ts [hunk 1]
```diff
@@ -137,6 +137,32 @@ describe('VcdNumberWithUnitFormInputComponent', () => {
             numberWithUnitInput.clickUnlimitedCheckbox();
             expect(numberWithUnitInput.isInputValueFocused).toBe(true, 'Input element should have focus');
         });
+
+        it('will choose a unit if initially unlimited when unchecked', () => {
```
Comment (irahov) [inline projects/components/src/form/number-with-unit-input/number-with-unit-form-input.component.spec.ts:141]: not clear isn't it better
will choose a unit when unchecked for initial value unlimited

File: projects/components/src/form/number-with-unit-input/number-with-unit-form-input.component.ts [hunk 2]
```diff
@@ -239,6 +239,10 @@ export class NumberWithUnitFormInputComponent extends BaseFormControl implements
         this.unlimitedControlValue = unlimitedChecked;
         this.textInputValue = unlimitedChecked ? '' : this.lastRealValue?.toString() || '';
 
+        if (!this.unitsControlValue && !unlimitedChecked) {
```
Comment (irahov) [inline projects/components/src/form/number-with-unit-input/number-with-unit-form-input.component.ts:242]: Some comment will help understand why in this case we should set the value by getting the value

File: projects/components/src/form/number-with-unit-input/number-with-unit-form-input.component.ts [hunk 3]
```diff
@@ -333,6 +337,15 @@ export class NumberWithUnitFormInputComponent extends BaseFormControl implements
         }
     }
 
+    /**
+     * Checks if the value in the form is the unlimited value.
```
Comment (irahov) [inline projects/components/src/form/number-with-unit-input/number-with-unit-form-input.component.ts:341]: It is confusing:
`Checks if the value in the form is the unlimited value` this means the method should return `true` or `false`

File: projects/examples/src/components/form-input/number-with-unit-form-input.example.component.ts [hunk 4]
```diff
@@ -52,8 +52,8 @@ export class NumberWithUnitFormInputExampleComponent implements AfterViewInit, O
         this.formGroup = fb.group({
             readonly: new FormControl(false),
             disabled: new FormControl(false),
-            memory: new FormControl(1024 * 2, [Validators.required, memoryValidator]),
```
Comment (irahov) [inline projects/examples/src/components/form-input/number-with-unit-form-input.example.component.ts:55]: Why did you remove the `memoryValidator` ?

File: projects/examples/src/components/form-input/number-with-unit-form-input.example.component.ts [hunk 5]
```diff
@@ -67,9 +67,5 @@ export class NumberWithUnitFormInputExampleComponent implements AfterViewInit, O
         });
     }
 
-    ngAfterViewInit(): void {
-        this.formGroup.controls.cpuLimit.setValue(-1);
```
Comment (irahov) [inline projects/examples/src/components/form-input/number-with-unit-form-input.example.component.ts:71]: I think there is a case to set the value later and not as part of the formGroup creation. But it is pitty there is no comment (may be it was me that did it and missed to add a comment).
Just double check please

**Review by irahov — APPROVED** [review]

File: projects/examples/src/components/form-input/number-with-unit-form-input.example.component.ts (hunk 5 above)
Comment (ryan-bradford) [reply projects/examples/src/components/form-input/number-with-unit-form-input.example.component.ts:71]: Ok. I'm going to merge this because Bryan needs the fixes. But we can talk about this on Monday in the parking lot.
